File: Shuffle-an-Array/SAA.go
package main

import (
	"fmt"
	"math/rand"
)

type Solution struct {
	arr map[int]int
}

func Constructor(nums []int) Solution {
	nullm := map[int]int{}
	a := Solution{arr: nullm}
	for i, e := range nums {
		a.arr[i] = e
	}

	return a
}

/** Resets the array to its original configuration and return it. */
func (this *Solution) Reset() []int {
	var a []int
	for i := 0; i < len(this.arr); i++ {
		a = append(a, this.arr[i])
	}

	return a
}

/** Returns a random shuffling of the array. */
func (this *Solution) Shuffle() []int {
	randArr := rand.Perm(len(this.arr))
	var a []int
	for _, i := range randArr {
		a = append(a, this.arr[i])
	}
	return a
}

func main() {
	nums := []int{1, 2, 3}

	obj := Constructor(nums)
	param_1 := obj.Reset()
	param_2 := obj.Shuffle()

	fmt.Println(param_1)
	fmt.Println(param_2)
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.Reset();
 * param_2 := obj.Shuffle();
 */

File: Shuffle-an-Array/SAA.swift
import Foundation

// Random extension function from http://stackoverflow.com/questions/37843647/shuffle-array-swift-3
extension Array {
    mutating func shuffle () {
        for i in (0..<self.count).reversed() {
            let ix1 = i
            let ix2 = Int(arc4random_uniform(UInt32(i+1)))
            (self[ix1], self[ix2]) = (self[ix2], self[ix1])
        }
    }
}


class Solution {
    var arr:[Int:Int] = [:]
    var ar:[Int]
    
    init(_ nums:[Int]) {
        self.ar = nums
        for (i,e) in nums.enumerated(){
            self.arr[i] = e
        }
    }

    func Reset() {
        for i in 0..<arr.count {
            self.ar[i] = self.arr[i]!
        }
    }

    func Shuffle() {
        self.ar.shuffle()
    }
    
}

var obj = Solution([1,2,3])
obj.Shuffle()
print(obj.ar)
obj.Reset()
print(obj.ar)
obj.Shuffle()
print(obj.ar)
